[Salesforce] Improve request handling

navarone-feekery opened the pull request:
## Related to https://github.com/elastic/enterprise-search-team/issues/5441

General improvements to the API request/response logic

- Add exception handling
- Add token refresh (not really a refresh)
- Add pagination handling

## Checklists

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [x] Considered corresponding documentation changes
- [x] Contributed any configuration settings changes to the configuration reference

artem-shelkovnikov commented on connectors/sources/salesforce.py:
```diff
@@ -80,58 +110,66 @@ async def close(self):
-        self.token_issued_at = resp_json["issued_at"]
-        self.token_refresh_enabled = enable_refresh
-        self._logger.debug("Salesforce token retrieved.")
+    async def get_token(self):
```
I'd probably move token logic into separate class.

Also, it might make sense to wrap the logic in a mutex.

That also gave me a hint, that we're missing this logic in Sharepoint Online connector!

navarone-feekery replied:
mutex + separate class sounds really clean, I like it! Thanks!

artem-shelkovnikov commented on connectors/sources/salesforce.py:
```diff
@@ -80,58 +110,66 @@ async def close(self):
+        except ClientResponseError as e:
+            if 400 <= e.status < 500:
+                # 400s have detailed error messages in body
+                error_resp = self._handle_token_response_body_error(response_body)
```
Is `e.message` different than response_body?

navarone-feekery replied:
Yeah they are different, for example if the credentials are invalid it throws a 400 with the message a `bad_request`. The actual cause (`invalid_client`) is then inside the response body.

artem-shelkovnikov commented on connectors/sources/salesforce.py:
```diff
@@ -169,26 +206,118 @@ async def _select_queryable_fields(self, sobject, fields):
+                page_count += 1
+                if (
+                    next_page.get("done") is True
+                    or page_count >= QUERY_PAGINATION_LIMIT
```
I believe usage of `nextRecordsUrl` eliminates the need to have QUERY_PAGINATION_LIMIT? I thought it applies only to skip/take pagination

navarone-feekery replied:
I just didn't want an endless loop, in case _something_ went wrong. It's really just a safeguard.

artem-shelkovnikov commented on connectors/sources/salesforce.py:
```diff
@@ -169,26 +206,118 @@ async def _select_queryable_fields(self, sobject, fields):
 
-    async def get_request(self, endpoint, params=None):
-        # TODO improve error handling
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+    )
```
I think you'll need custom retryable logic here - no need to retry on invalid query, probably on throttling too

artem-shelkovnikov commented on connectors/sources/salesforce.py:
```diff
@@ -169,26 +206,118 @@ async def _select_queryable_fields(self, sobject, fields):
+            response = await self._get(url, params=params)
+            response_body = await response.json()
+            # We get the response body before raising for status as it contains vital error information
+            response.raise_for_status()
```
Hmmm, I haven't used this method before - I thought response will raise by default? Or is it only if you use it in context manager style?

navarone-feekery replied:
If you set `raise_for_status=True` in either the session or the request, then it will raise (it's `False` by default)

artem-shelkovnikov commented on connectors/sources/salesforce.py:
```diff
@@ -169,26 +206,118 @@ async def _select_queryable_fields(self, sobject, fields):
+            )
+            # The user can alter the lifetime of issued tokens, so we don't know when they expire
+            # Therefore we fetch the token when we encounter an error rather than when it expires
+            await self.get_token()
```
Might it be worth to actually wipe the token before fetching?

wangch079 commented on connectors/sources/salesforce.py:
```diff
@@ -35,13 +37,42 @@
 
 
+class RateLimitedException(Exception):
+    """Notifies that Salesforce has begun rate limiting the current accound"""
```
```suggestion
    """Notifies that Salesforce has begun rate limiting the current account"""
```

wangch079 commented on connectors/sources/salesforce.py:
```diff
@@ -169,26 +206,118 @@ async def _select_queryable_fields(self, sobject, fields):
-            QUERY_ENDPOINT,
+    async def _yield_non_bulk_query_pages(self, soql_query):
+        """loops through query response pages and yields lists of records"""
+        response = await self._get_json(
+            f"{self.base_url}{QUERY_ENDPOINT}",
             params={"q": soql_query},
         )
+        yield response.get("records")
```
Why not use `while True` and `break`?

wangch079 commented on connectors/sources/salesforce.py:
```diff
@@ -169,26 +206,118 @@ async def _select_queryable_fields(self, sobject, fields):
+                )
+                in error_codes
+            ):
+                raise InvalidQueryException(
```
If it's `RateLimitedException` or `InvalidQueryException`, we should not retry

wangch079 replied:
You can use this when it's merged: https://github.com/elastic/connectors-python/pull/1395

navarone-feekery replied:
That was a really convenient update, thank you for adding it!

navarone-feekery commented:
@artem-shelkovnikov @wangch079 I think I've addressed your comments. I need to merge this to unblock another PR, so if there's anything else that you think requires changing let me know and I'll make a followup PR 🙂
